wd_helper.au3

```diff
@@ -538,48 +538,78 @@ EndFunc   ;==>_WD_IsWindowTop
 ; Description ...: Enter the specified frame.
 ; Syntax ........: _WD_FrameEnter($sSession, $vIdentifier)
 ; Parameters ....: $sSession    - Session ID from _WD_CreateSession
-;                  $vIdentifier - Index (as 0-based Integer) or Element ID (as String) or Null (Keyword)
+;                  $vIdentifier - Target frame identifier. Can be any of the following:
+;                  |Null    - Return to top-most browsing context
+;                  |String  - Element ID from _WD_FindElement or path like 'null/2/0'
+;                  |Integer - 0-based index of frames
 ; Return values .: Success - True.
-;                  Failure - WD Response error message (E.g. "no such frame") and sets @error to $_WD_ERROR_Exception
+;                  Failure - WD Response error message (E.g. "no such frame") and sets @error to $_WD_ERROR_Exception or $_WD_ERROR_InvalidArgue
 ; Author ........: Decibel
 ; Modified ......: mLipok
-; Remarks .......: You can drill-down into nested frames by calling this function repeatedly with the correct parameters.
+; Remarks .......: You can drill-down into nested frames by calling this function repeatedly with the correct parameters or use path like 'null/2/0'
 ; Related .......: _WD_Window, _WD_LastHTTPResult
 ; Link ..........:
 ; Example .......: No
 ; ===============================================================================================================================
 Func _WD_FrameEnter($sSession, $vIdentifier)
 	Local Const $sFuncName = "_WD_FrameEnter"
-	Local Const $sParameters = 'Parameters:    Identifier=' & $vIdentifier
-	Local $sOption, $sValue, $sResponse, $oJSON
+	If String($vIdentifier) = 'null' Then $vIdentifier = Null ; String must be used because checking 0 = 'null' is True
+	Local Const $bIsIdentifierNull = (IsKeyword($vIdentifier) = $KEYWORD_NULL)
+	Local Const $sParameters = 'Parameters:    Identifier=' & ($bIsIdentifierNull ? ("Null") : ($vIdentifier))
+	Local $sValue, $sMessage = '', $sOption, $sResponse, $oJSON
 	Local $iErr = $_WD_ERROR_Success
 
 	;*** Encapsulate the value if it's an integer, assuming that it's supposed to be an Index, not ID attrib value.
-	If (IsKeyword($vIdentifier) = $KEYWORD_NULL) Then
+	Local Const $aIdentifiers = StringSplit($vIdentifier, '/')
+	Local Const $bIdentifierAsPath = ($aIdentifiers[0] > 1)
+	If $bIdentifierAsPath Then
+		For $i = 1 To $aIdentifiers[0]
+			If String($aIdentifiers[$i]) = 'null' Then
+				$sValue = _WD_FrameEnter($sSession, Null)
+				$iErr = @error
+			ElseIf StringIsDigit($aIdentifiers[$i]) And IsInt(Number($aIdentifiers[$i])) Then
+				$sValue = _WD_FrameEnter($sSession, Int($aIdentifiers[$i]))
+				$iErr = @error
```

Comment:
IMO it be better to call _WD_Window here instead of recursively calling _WD_FrameEnter.

Reply:
did you also think about all this code 
```autoit
	If $iErr = $_WD_ERROR_Success And Not $bIdentifierAsPath Then ; check if $vIdentifier was succesfully validated and not given "as path" with /
		$sResponse = _WD_Window($sSession, "frame", $sOption)
		$iErr = @error

		If $iErr = $_WD_ERROR_Success Then
			$oJSON = Json_Decode($sResponse)
			$sValue = Json_Get($oJSON, $_WD_JSON_Value)

			;*** Evaluate the response
			If $sValue <> Null Then
				$sValue = Json_Get($oJSON, $_WD_JSON_Error)
			Else
				$sValue = True
			EndIf
		ElseIf $iErr <> $_WD_ERROR_InvalidArgue Then
			$iErr = $_WD_ERROR_Exception
		EndIf
	EndIf
```

that should be repeated ?

Reply:
The area of code I referenced was inside the `If $bIdentifierAsPath Then`, so the code you linked above wouldn't apply, right?

Reply:
I see solution.
I will implement your idea.

Reply:
Please take a look on: [no more recursively calling _WD_FrameEnter](https://github.com/Danp2/au3WebDriver/pull/364/commits/c183e5c9955fc5d0c3a3401773a7611ea7dc9468)

Reply:
I thought that we had decided that "0/1" was valid for `$vIdentifier`. If that's correct, then your Regex won't work.

Reply:
ah sorry. should this be fine `(?i)\A(Null\/|)([\d\/]*\d)\Z` ?

Reply:
This appears to fix one issue, but creates another.

Reply:
What kind of another issues ?

Reply:
In my testing, it included the zero-based integer indexes, which is wrong. Does this match your results?

Reply:
Using this:
```autoit
Func UserTesting() ; here you can replace the code to test your stuff before you ask on the forum
	$_WD_DEBUG = $_WD_DEBUG_Error
	_WD_Navigate($sSession, 'https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe')
	_WD_LoadWait($sSession)

	$_WD_DEBUG = $_WD_DEBUG_Full
	_WD_FrameEnter($sSession, 'null/2')
	MsgBox(0, "", 'test1')
	_WD_FrameEnter($sSession, '0/0')
	MsgBox(0, "", 'test2')

	$_WD_DEBUG = $_WD_DEBUG_Error
	; Example 2 - from 'https://www.w3schools.com' get frame list as array, with HTML content of each document
	Local $aFrameList = _WD_FrameList($sSession, True, '', True)
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & " : Example 2" & @CRLF)
	_ArrayDisplay($aFrameList, 'Example 2', 0, 0, Default, 'Absolute Identifiers > _WD_FrameEnter|Relative Identifiers > _WD_FrameEnter|IFRAME attributes|URL|Body ElementID|HTML')
EndFunc
```
I got
```
+ wd_demo.au3: Running: UserTesting
__WD_Post: URL=HTTP://127.0.0.1:4444/session/9cc4030f-be99-4021-b591-658f73e2ffb3/frame; Data={"id":null}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":null}
__WD_Post: URL=HTTP://127.0.0.1:4444/session/9cc4030f-be99-4021-b591-658f73e2ffb3/frame; Data={"id":2}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":2}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=null/2
__WD_Post: URL=HTTP://127.0.0.1:4444/session/9cc4030f-be99-4021-b591-658f73e2ffb3/frame; Data={"id":0}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":0}
__WD_Post: URL=HTTP://127.0.0.1:4444/session/9cc4030f-be99-4021-b591-658f73e2ffb3/frame; Data={"id":0}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":0}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=0/0
! ---> @error=0  @extended=0 : Example 2
```

and
![image](https://user-images.githubusercontent.com/11089482/183298276-c1187c2e-c5ea-4161-8733-d2673f1cae8e.png)


So I'm not sure what you mean.
Can you post your `UserTesting()` function ?

Reply:
There are scenarios that you failed to test AFAICS. Try both of these --

```autoit
	_WD_FrameEnter($sSession, Null)
	_WD_FrameEnter($sSession, 0)
```

Do they both work correctly?

Reply:
using:
```autoit
Func UserTesting() ; here you can replace the code to test your stuff before you ask on the forum
	$_WD_DEBUG = $_WD_DEBUG_Error
	_WD_Navigate($sSession, 'https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe')
	_WD_LoadWait($sSession)

	$_WD_DEBUG = $_WD_DEBUG_Full
	_WD_FrameEnter($sSession, Null)
	MsgBox(0, "", 'test1')
	_WD_FrameEnter($sSession, 0)
	MsgBox(0, "", 'test2')

	$_WD_DEBUG = $_WD_DEBUG_Error
	; Example 2 - from 'https://www.w3schools.com' get frame list as array, with HTML content of each document
	Local $aFrameList = _WD_FrameList($sSession, True, '', True)
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & " : Example 2" & @CRLF)
	_ArrayDisplay($aFrameList, 'Example 2', 0, 0, Default, 'Absolute Identifiers > _WD_FrameEnter|Relative Identifiers > _WD_FrameEnter|IFRAME attributes|URL|Body ElementID|HTML')
EndFunc
```

I get:
```
+ wd_demo.au3: Running: UserTesting
__WD_Post: URL=HTTP://127.0.0.1:4444/session/29adb83a-f422-4247-8ba3-4b26d7b1aa1f/frame; Data={"id":null}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":null}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=Null
__WD_Post: URL=HTTP://127.0.0.1:4444/session/29adb83a-f422-4247-8ba3-4b26d7b1aa1f/frame; Data={"id":0}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":0}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=0
! ---> @error=0  @extended=0 : Example 2
```

![image](https://user-images.githubusercontent.com/11089482/183308153-a364d082-a46d-46c0-82d0-9a1f22c30494.png)

Reply:
To be sure you have added

> ah sorry. should this be fine `(?i)\A(Null\/|)([\d\/]*\d)\Z` ?

I just commit [$bIdentifierAsPath StringRegExp fixed](https://github.com/Danp2/au3WebDriver/pull/364/commits/15926559214b49e5fdc9332ba4691c0b5bbf548e)

This is exactly the same version which I'm using currently.

Please try it. Hope it solves the issue on your side.

Reply:
The code runs fine. However, the issue is that your regex captures values like 0, 1, etc, which is what I tried to explain when I said "In my testing, it included the zero-based integer indexes, which is wrong."

With the line `	_WD_FrameEnter($sSession, 0)`, `$bIdentifierAsPath` evaluates to non-zero and this section of code will never run --

```autoit
ElseIf StringIsDigit($vIdentifier) And IsInt(Number($vIdentifier)) Then
		$sOption = '{"id":' & $vIdentifier & '}'
```

I suspect that this was an unintended consequence of changing the regex.

Reply:
Oh. Now I understand.
Will check and fix it ASAP

Reply:
I find that `(?i)\A(Null|\d)(\/)([\d\/]*\d)\Z` fix the issue, but I notice another possible issue: such string `Null/1/2121/////2111` will also be validated as true.
I will ask for help JCHD

Reply:
Please take a look on: [$bIdentifierAsPath fix](https://github.com/Danp2/au3WebDriver/pull/364/commits/af42b5cdca7b1246fa1a83042159aedc4cd0a34e)

Reply:
`jchd` solution:
```autoit
#include <StringConstants.au3>
_Example()
Exit

Func _Example()
	Local $a = ["valid forms", "Null/1/2121/2", "null/0", "1/2/3", "1/2", "now invalid forms", "", "Null/1/2121/////2111", "true/1/2121/////2111", "abcd/1/2121/2", "1/", "/1", "1", 1, "-1", -1, Null]
	For $s In $a
		ConsoleWrite($s & @TAB & StringRegExp($s, "(?i)\A(?:Null|\d+)(?:\/\d+)+\Z", $STR_REGEXPMATCH) & @LF)
	Next
EndFunc
```